File: source/schemas/Message.go
package schemas

type Message struct {
	Role             string     `json:"role" yaml:"role"`
	Content          string     `json:"content" yaml:"content"`
	ReasoningContent string     `json:"reasoning_content,omitempty" yaml:"reasoning-content,omitempty"`
	ToolCallID       string     `json:"tool_call_id,omitempty" yaml:"tool-call-id,omitempty"`
	ToolCalls        []ToolCall `json:"tool_calls,omitempty" yaml:"tool-calls,omitempty"`
	ToolName         string     `json:"tool_name,omitempty" yaml:"tool-name,omitempty"`
	Created          Datetime   `json:"created,omitempty" yaml:"created,omitempty"`
}

File: source/ui/web/public/schemas/Message.mjs

import { ToolCall } from "/schemas/ToolCall.mjs";

export const Message = function() {

	this.Role       = "";
	this.Content    = "";
	this.Created    = new Date();
	this.ToolCallID = "";
	this.ToolName   = "";
	this.ToolCalls  = [];

};

Message.from = (data) => {

	let message = new Message();

	message.Role       = data["role"]    || "";
	message.Content    = data["content"] || "";
	message.Created    = new Date((data["created"] || "").replace(" ", "T"));
	message.ToolCallID = data["tool_call_id"] || "";
	message.ToolName   = data["tool_name"]    || "";

	if (Array.isArray(data["tool_calls"]) === true) {

		message.ToolCalls = data["tool_calls"].map((raw) => {
			return ToolCall.from(raw);
		});

	}

	return message;

};
